lib/test_state.go

```diff
@@ -0,0 +1,34 @@
+package lib
+
+import (
+	"io"
+
+	"github.com/sirupsen/logrus"
+	"go.k6.io/k6/metrics"
+)
+
+// TestPreInitState contains all of the state that can be gathered and built
+// before the test run is initialized.
+type TestPreInitState struct {
```

Comment:
I regret that the name is somewhat implementation specific (`PreInit` refers to "happens before `ExecutionScheduler.Init` is called right?), and would rather have it being something more specific to the commands themselves: we're trying to get k6 to perform some action, and this is the state we start in, before initializing any k6 components. Maybe `TestInitialState`, and documenting the context would do?

 ```suggestion
// TestRunInitialState contains the state gathered and consolidated before any k6 components were initialized.
//
// This state is useful because of A, B, C.
// This state is used in scenarios X, Y, Z.
// Use this state when you want to achieve "foo".
type TestRunInitialState struct {
    // RuntimeOptions holds the "runtime" configuration we want to pass into the test script (no-thresholds, etc...)
    RuntimeOptions RuntimeOptions
    
    // etc... 
}
```

I also noticed, as you mentioned in another comment, that we have somewhat of a mix of state and configuration in there too 👍🏻

Reply:
Pre-init in the sense of https://k6.io/docs/using-k6/test-life-cycle, so it's just implementation specific. We have documented the init phase as the initial execution k6 does of the global scope, to get the exported script `options` and functions and such. So this struct basically contains most of the state and options we have gathered before we can execute this init phase.

Reply:
FWIW, I am not super happy with either `TestPreInitState` or `TestRunState` :sweat_smile: The nice thing is that, if we can figure out better names in the future, that will be an easy refactor to do, since their contents and method signatures won't change :tada: